coderd/notifications_test.go

```diff
@@ -0,0 +1,79 @@
+package coderd_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/coder/coder/v2/coderd/coderdtest"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/testutil"
+)
+
+func TestUpdateNotificationsSettings(t *testing.T) {
+	t.Parallel()
+
+	t.Run("Permissions denied", func(t *testing.T) {
+		t.Parallel()
+
+		api := coderdtest.New(t, nil)
+		firstUser := coderdtest.CreateFirstUser(t, api)
+		anotherClient, _ := coderdtest.CreateAnotherUser(t, api, firstUser.OrganizationID)
+
+		// given
+		expected := codersdk.NotificationsSettings{
+			NotifierPaused: true,
+		}
+
+		ctx := testutil.Context(t, testutil.WaitShort)
+
+		// when
+		err := anotherClient.PutNotificationsSettings(ctx, expected)
+
+		// then
+		require.Error(t, err) // Insufficient permissions to update notifications settings.
+	})
+
+	t.Run("Settings modified", func(t *testing.T) {
+		t.Parallel()
+
+		client := coderdtest.New(t, nil)
+		_ = coderdtest.CreateFirstUser(t, client)
+
+		// given
+		expected := codersdk.NotificationsSettings{
+			NotifierPaused: true,
+		}
+
+		ctx := testutil.Context(t, testutil.WaitShort)
+
+		// when
+		err := client.PutNotificationsSettings(ctx, expected)
+		require.NoError(t, err)
+
+		// then
+		actual, err := client.GetNotificationsSettings(ctx)
+		require.NoError(t, err)
+		require.Equal(t, expected, actual)
+	})
+
+	t.Run("Settings not modified", func(t *testing.T) {
+		t.Parallel()
+
+		client := coderdtest.New(t, nil)
+		_ = coderdtest.CreateFirstUser(t, client)
+
+		ctx := testutil.Context(t, testutil.WaitShort)
+
+		// given
+		expected := codersdk.NotificationsSettings{
+			NotifierPaused: false,
+		}
+		err := client.PutNotificationsSettings(ctx, expected)
+		require.NoError(t, err)
+
+		// then
+		err = client.PutNotificationsSettings(ctx, expected)
+		require.Error(t, err) // Error: notifications settings not modified
```

Comment:
Ditto

Reply:
Why should another call to `PutNotificationsSettings` with the same contents return an error?

Reply:
This way we can indicate that it was HTTP 304 (see below). It is a common pattern in Coder client?

```
	if res.StatusCode == http.StatusNotModified {
		return xerrors.New("notifications settings not modified")
	}
```

Reply:
> Even so, we should check for the very specific outcome we want, not just for any error.

Ok, I will just return `nil` and won't indicate an error

Reply:
Fixed